sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/OrderedEventProcessor.java

```diff
@@ -0,0 +1,657 @@
+          window.maxTimestamp());
+
+      checkIfProcessingIsCompleted(processingState);
+    }
+
+    private boolean checkIfProcessingIsCompleted(ProcessingState<EventKeyTypeT> processingState) {
+      boolean result = processingState.isProcessingCompleted();
+      if (result) {
+        LOG.info("Processing for key '" + processingState.getKey() + "' is completed.");
+      }
+      return result;
+    }
+
+    private void saveStates(
+        ValueState<ProcessingState<EventKeyTypeT>> processingStatusState,
+        ProcessingState<EventKeyTypeT> processingStatus,
+        ValueState<StateTypeT> currentStateState,
+        StateTypeT state,
+        MultiOutputReceiver outputReceiver,
+        Instant windowTimestamp) {
+      // There is always a change to the processing status
+      processingStatusState.write(processingStatus);
+
+      // Stored state may not have changes if the element was out of sequence.
+      if (state != null) {
+        currentStateState.write(state);
+      }
+
+      if (produceStatusUpdateOnEveryEvent) {
+        // During pipeline draining the window timestamp is set to a large value in the future.
+        // Producing an event before that results in error, that's why this logic exist.
+        Instant statusTimestamp = Instant.now();
+        statusTimestamp =
+            statusTimestamp.isAfter(windowTimestamp) ? statusTimestamp : windowTimestamp;
+
+        emitProcessingStatus(processingStatus, outputReceiver, statusTimestamp);
+      }
+    }
+
+    private void emitProcessingStatus(
+        ProcessingState<EventKeyTypeT> processingState,
+        MultiOutputReceiver outputReceiver,
+        Instant statusTimestamp) {
+      outputReceiver
+          .get(statusTupleTag)
+          .outputWithTimestamp(
+              KV.of(
+                  processingState.getKey(),
+                  OrderedProcessingStatus.create(
+                      processingState.getLastOutputSequence(),
+                      processingState.getBufferedRecordCount(),
+                      processingState.getEarliestBufferedSequence(),
+                      processingState.getLatestBufferedSequence(),
+                      processingState.getRecordsReceived(),
+                      processingState.getResultCount(),
+                      processingState.getDuplicates(),
+                      processingState.isLastEventReceived())),
+              statusTimestamp);
+    }
+
+    /**
+     * Process the just received event.
+     *
+     * @return newly created or updated State. If null is returned - the event wasn't processed.
+     */
+    private StateTypeT processNewEvent(
+        long currentSequence,
+        EventTypeT currentEvent,
+        ProcessingState<EventKeyTypeT> processingState,
+        ValueState<StateTypeT> currentStateState,
+        OrderedListState<EventTypeT> bufferedEventsState,
+        MultiOutputReceiver outputReceiver) {
+      if (currentSequence == Long.MAX_VALUE) {
+        LOG.error(
```

Comment:
Rather than logging and dropping, maybe lets throw an exception here? Seems like this is always a user code error and could lead to lost data

Reply:
If we do that, the pipeline will have to be cancelled - it won't drain properly because the event is unprocessable. That could lead to larger data loss (all the events currently in processing). Thoughts?

Reply:
That's a good point. With that said, as written this data is unrecoverable (there's no identifying info other than the sequence number).

I guess this is low enough risk to leave as is though; a MAX_VALUE sequence number is very unlikely to be passed in anyways, and I think somewhat non-deterministic behavior around that user error is probably ok

Reply:
I changed the logic to output this event to the DLQ PCollection with the specific failure reason. This way there is no data loss.